### PR TITLE
PF-1965: When cloning resources, clear certain properties (a.k.a TERRA_WORKSPACE_FOLDER_ID)

### PR DESCRIPTION
**Pull request by yuhuyoyo**
original gcs-bucket:
```
{
  "resources": [
    {
      "metadata": {
        "workspaceId": "d6765c27-93a3-4573-bc4c-9e509f080a95",
        "resourceId": "0c737261-874b-4c02-8d86-d00843796aa3",
        "name": "my-bucket",
        "resourceType": "GCS_BUCKET",
        "stewardshipType": "CONTROLLED",
        "cloudPlatform": "GCP",
        "cloningInstructions": "COPY_RESOURCE",
        "controlledResourceMetadata": {
          "accessScope": "SHARED_ACCESS",
          "managedBy": "USER",
          "privateResourceUser": {},
          "privateResourceState": "NOT_APPLICABLE"
        },
        "resourceLineage": [],
        "properties": [
          {
            "key": "TERRA_WORKSPACE_FOLDER_ID",
            "value": "6b04f199-d895-4d7e-ab6f-1d0bd64818af"
          }
        ]
      },
      "resourceAttributes": {
        "gcpGcsBucket": {
          "bucketName": "my-bucket-terra-wsm-t-swift-turnip-3879"
        }
      }
    }
  ]
}
```

Cloned gcs-bucket:
```
{
  "resources": [
    {
      "metadata": {
        "workspaceId": "4d1eb726-0bf6-49a7-91d9-067b18ee8585",
        "resourceId": "48751596-9263-44f1-8d70-afb51b0c8d0a",
        "name": "my-bucket",
        "resourceType": "GCS_BUCKET",
        "stewardshipType": "CONTROLLED",
        "cloudPlatform": "GCP",
        "cloningInstructions": "COPY_RESOURCE",
        "controlledResourceMetadata": {
          "accessScope": "SHARED_ACCESS",
          "managedBy": "USER",
          "privateResourceUser": {},
          "privateResourceState": "NOT_APPLICABLE"
        },
        "resourceLineage": [
          {
            "sourceWorkspaceId": "d6765c27-93a3-4573-bc4c-9e509f080a95",
            "sourceResourceId": "0c737261-874b-4c02-8d86-d00843796aa3"
          }
        ],
        "properties": []
      },
      "resourceAttributes": {
        "gcpGcsBucket": {
          "bucketName": "cloned-my-bucket-terra-wsm-t-bright-corn-4451"
        }
      }
    }
  ]
}
```

### CODE REVIEW

File: service/src/main/java/bio/terra/workspace/app/controller/shared/PropertiesUtils.java [hunk 1]
```diff
@@ -28,4 +34,16 @@ public static ApiProperties convertMapToApiProperties(Map<String, String> proper
...
     return apiProperties;
   }
+
+  /**
+   * Clear certain properties in the hashmap before making a clone.
+   *
+   * <p>For example, TERRA_WORKSPACE_FOLDER_ID is a workspace specific properties. It needs to be
+   * cleared because it is meaningless in a new workspace.
+   */
+  public static ImmutableMap<String, String> clearSomeProperties(Map<String, String> properties) {
```
Comment (melissachang) [inline service/src/main/java/bio/terra/workspace/app/controller/shared/PropertiesUtils.java:44]: UI question. Ignore Data Collection UI.

- Say in UI resource R is in folder F in workspace W
- Say I clone R into same workspace W
- By default, I would think the clone would be in the same folder F. Eg in google docs, if I make a copy, the copy is in the same folder.

I wonder if we should keep folder if source and dest workspace are the same.

@cbookg thoughts?

File: service/src/main/java/bio/terra/workspace/app/controller/shared/PropertiesUtils.java [hunk 2]
```diff
@@ -28,4 +34,16 @@ public static ApiProperties convertMapToApiProperties(Map<String, String> proper
     properties.forEach((key, value) -> apiProperties.add(new ApiProperty().key(key).value(value)));
     return apiProperties;
   }
+
+  /**
+   * Clear certain properties in the hashmap before making a clone.
```
Comment (melissachang) [inline service/src/main/java/bio/terra/workspace/app/controller/shared/PropertiesUtils.java:39]: before cloning a resource

(clarifying this is for resource clone, not workspace clone)

**Review by melissachang — APPROVED** [review]

File: service/src/main/java/bio/terra/workspace/app/controller/shared/PropertiesUtils.java (hunk 1 above)
Comment (zloery) [inline service/src/main/java/bio/terra/workspace/app/controller/shared/PropertiesUtils.java:44]: this method name should indicate this is clearing properties relevant to cloning.
it's also slightly confusing to name it `clear___` given that it's really returning a copy without modifying the input. 

maybe something like `clonableProperties` or `getClonableProperties` instead?

**Review by zloery — APPROVED** [review]

File: service/src/main/java/bio/terra/workspace/app/controller/shared/PropertiesUtils.java (hunk 1 above)
Comment (yuhuyoyo) [reply service/src/main/java/bio/terra/workspace/app/controller/shared/PropertiesUtils.java:44]: it does modify the input. i modified comment and method name to make it clearer.

File: service/src/test/java/bio/terra/workspace/service/resource/controlled/flight/clone/workspace/WorkspaceCloneUtilsTest.java [hunk 3]
```diff
@@ -81,6 +88,30 @@ public void buildDestinationControlledBigQueryDataset_cloneSucceeds() {
     assertControlledResourceCommonField(sourceDataset, datasetToClone);
   }
 
+  @Test
```
[the remark this answers is not in the document]
Comment (yuhuyoyo) [reply service/src/test/java/bio/terra/workspace/service/resource/controlled/flight/clone/workspace/WorkspaceCloneUtilsTest.java:91]: i like to add test to A when i change code in A.

File: service/src/main/java/bio/terra/workspace/app/controller/shared/PropertiesUtils.java (hunk 1 above)
Comment (yuhuyoyo) [reply service/src/main/java/bio/terra/workspace/app/controller/shared/PropertiesUtils.java:44]: good point. i think that makes sense to keep the folder properties if cloning into the same workspace.

File: service/src/main/java/bio/terra/workspace/app/controller/shared/PropertiesUtils.java [hunk 4]
```diff
@@ -11,6 +13,10 @@
 /** Utils for properties that are key value pairs in WSM workspace and resources. */
 public class PropertiesUtils {
 
+  public static class ResourcePropertiesKey {
+    public static final String FOLDER_ID_KEY = "TERRA_WORKSPACE_FOLDER_ID";
```
[the remark this answers is not in the document]
Comment (yuhuyoyo) [reply service/src/main/java/bio/terra/workspace/app/controller/shared/PropertiesUtils.java:17]: fine with me.. i changed it.

File: service/src/main/java/bio/terra/workspace/app/controller/shared/PropertiesUtils.java (hunk 1 above)
Comment (yuhuyoyo) [reply service/src/main/java/bio/terra/workspace/app/controller/shared/PropertiesUtils.java]: i modified it to keep folder if source and dest workspaceid are the same

**Pull request merged**